### PR TITLE
Provider updates

### PR DESCRIPTION
**Pull request by robmorgan**
This PR updates the repo to support the latest versions of TF 0.12 and the Google provider. It also is a fairly major refactor that includes the following updates:

- [x] Updated Consul to 1.8.3.
- [x] Added support for rolling cluster updates using an update policy.
- [x] Dropped support for Ubuntu 16.04 LTS.
- [x] Support for Google provider 3.38.0.
- [x] Dropped Dep in favor of Go modules.
- [x] Support for Packer 1.6.x.

### CODE REVIEW

**Comment by robmorgan** [comment]
@yorinasub17 for some reason I can't add you to review this PR?!

**Comment by yorinasub17** [comment]
> for some reason I can't add you to review this PR?!

It's most likely because I don't have write access to this repo.

**Comment by robmorgan** [comment]
@yorinasub17 ahh, should we change that or should I chase a review from @brikis98 instead?

File: modules/consul-cluster/main.tf [hunk 1]
```diff
@@ -135,7 +89,21 @@ resource "google_compute_instance_template" "consul_server_private" {
...
     subnetwork         = var.subnetwork_name != null ? var.subnetwork_name : null
-    subnetwork_project = var.network_project_id != null ? var.network_project_id : var.gcp_project_id
+    subnetwork_project = var.network_project_id != null ? var.network_project_id : null
+
+    #access_config {
+    # The presence of this property assigns a public IP address to each Compute Instance. We intentionally leave it
+    # blank so that an external IP address is selected automatically.
+    #  nat_ip = ""
+    #}
+
```
Comment (brikis98) [inline modules/consul-cluster/main.tf:99]: Intentional?

File: modules/install-consul/install-consul [hunk 2]
```diff
@@ -2,95 +2,93 @@
...
+
+  log_error "$description failed after 5 attempts."
+  exit $exit_status
 }
 
 function install_dependencies {
   log_info "Installing dependencies"
 
-  if $(has_apt_get); then
+  if $(os_is_ubuntu); then
```
Comment (brikis98) [inline modules/install-consul/install-consul:75]:
```suggestion
  if os_is_ubuntu; then
```

File: modules/install-consul/install-consul [hunk 3]
```diff
@@ -2,95 +2,93 @@
...
-function user_exists {
-  local -r username="$1"
-  id "$username" >/dev/null 2>&1
-}
-
 function create_consul_user {
   local -r username="$1"
 
-  if $(user_exists "$username"); then
+  if $(os_user_exists "$username"); then
```
Comment (brikis98) [inline modules/install-consul/install-consul:87]:
```suggestion
  if os_user_exists "$username"; then
```

File: modules/install-dnsmasq/install-dnsmasq [hunk 4]
```diff
@@ -33,49 +43,12 @@ function print_usage {
...
-  [[ -n "$(command -v apt-get)" ]]
-}
-
 function install_dnsmasq {
   local -r consul_ip="$1"
 
   log_info "Installing Dnsmasq"
 
-  if $(has_apt_get); then
+  if $(os_is_ubuntu); then
```
Comment (brikis98) [inline modules/install-dnsmasq/install-dnsmasq:51]:
```suggestion
  if os_is_ubuntu; then
```

File: modules/run-consul/run-consul [hunk 5]
```diff
@@ -123,53 +129,86 @@ function get_instance_ip_address {
...
   else
-    retry_join_key_val="\"retry_join\": [\"provider=gce project_name=$project_id tag_value=$cluster_tag_name\"],"
+    retry_join_json=$(cat <<EOF
+"retry_join": ["provider=gce project_name=$project_id tag_value=$cluster_tag_name"],
+EOF
+)
   fi
 
+  local recursors_config=""
+  if (( ${#recursors[@]} != 0 )); then
```
Comment (brikis98) [inline modules/run-consul/run-consul:191]:
```suggestion
  if [[ ${#recursors[@]} -ne 0 ]]; then
```

File: modules/run-consul/run-consul [hunk 6]
```diff
@@ -123,53 +129,86 @@ function get_instance_ip_address {
...
+        recursors_config="\"recursors\" : [ "
+        for recursor in "${recursors[@]}"
+        do
+            recursors_config="${recursors_config}\"${recursor}\", "
+        done
+        recursors_config=$(echo "${recursors_config}"| sed 's/, $//')" ],"
```
Comment (brikis98) [inline modules/run-consul/run-consul:197]: NIT: Wrong indentation?

File: modules/run-consul/run-consul [hunk 7]
```diff
@@ -182,58 +221,120 @@ function generate_consul_config {
...
-user=$consul_user
-# Per Supervisord docs (http://supervisord.org/configuration.html), when supervisord runs a process it does not start a
-# login shell and does not change environment variables like USER or HOME, so we must pass this in manually.
-environment=HOME="$consul_user_home_dir"
+  local -r consul_systemd_stdout="$4"
+  local -r consul_systemd_stderr="$5"
+  local -r consul_bin_dir="$6"
+  local -r consul_user="$7"
+  shift 7
+  local -r environment=("$@")
```
Comment (brikis98) [inline modules/run-consul/run-consul:279]:
```suggestion
  local -ar environment=("$@")
```

File: modules/run-consul/run-consul [hunk 8]
```diff
@@ -123,53 +129,86 @@ function get_instance_ip_address {
...
+  local -r max_trailing_logs="${16}"
+  local -r server_stabilization_time="${17}"
+  local -r redundancy_zone_tag="${18}"
+  local -r disable_upgrade_migration="${19}"
+  local -r upgrade_version_tag=${20}
   local -r config_path="$config_dir/$CONSUL_CONFIG_FILE"
 
-  local instance_ip_address=""
+  shift 20
+  local -r recursors=("$@")
```
Comment (brikis98) [inline modules/run-consul/run-consul:165]:
```suggestion
  local -ar recursors=("$@")
```

File: main.tf [hunk 9]
```diff
@@ -63,10 +64,15 @@ module "consul_servers" {
   # NOT for production usage. In production, set this to false.
   assign_public_ip_addresses = true
 
-  # WARNING! This update strategy will delete and re-create the entire Consul cluster when making some changes to this
-  # module. Unfortunately, Google and Terraform do not yet support an automatic stable way of performing a rolling update.
-  # For now for production usage, set this to "NONE", and manually coordinate your Consul Server upgrades per Consul docs.
-  instance_group_update_strategy = "NONE"
+  # This update strategy will performing a rolling update of the Consul cluster server nodes. We wait 5 minutes for
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply main.tf:67]: Yes 👍

File: main.tf [hunk 10]
```diff
@@ -110,16 +117,20 @@ module "consul_clients" {
...
 
   assign_public_ip_addresses = true
 
   source_image     = var.consul_client_source_image
   image_project_id = var.image_project_id
 
   # Our Consul Clients are completely stateless, so we are free to destroy and re-create them as needed.
-  # Todo: Research this further
-  instance_group_update_strategy = "NONE"
+  instance_group_update_policy_type                  = "PROACTIVE"
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply main.tf:128]: @brikis98 I think its actually okay: 
> Our Consul Clients are completely stateless, so we are free to destroy and re-create them as needed.

I keep the `var.instance_group_update_policy_min_ready_sec` value low and ensure all nodes can be 'unavailable' at the same time by setting `var.instance_group_update_policy_max_unavailable_fixed`, which makes for a faster deployment but also terminates the client nodes early without waiting for them to propagate as they are stateless anyway.

File: modules/consul-cluster/main.tf (hunk 1 above)
Comment (robmorgan) [reply modules/consul-cluster/main.tf:99]: Nope that needs to go, thanks for spotting!

File: .circleci/config.yml [hunk 11]
```diff
@@ -1,48 +1,57 @@
...
+      # The weird way you have to set PATH in Circle 2.0
+      - run: echo 'export PATH=$HOME/.local/bin:$HOME/terraform:$PATH' >> $BASH_ENV
+
+      - run:
+          name: install gruntwork tooling
+          command: |
+            sudo apt-get -y update
+            curl -Ls https://raw.githubusercontent.com/gruntwork-io/gruntwork-installer/master/bootstrap-gruntwork-installer.sh | bash /dev/stdin --version v0.0.21
+            gruntwork-install --module-name "gruntwork-module-circleci-helpers" --repo "https://github.com/gruntwork-io/module-ci" --tag "v0.16.6"
+            gruntwork-install --binary-name "terratest_log_parser" --repo "https://github.com/gruntwork-io/terratest" --tag "v0.13.9"
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply .circleci/config.yml:21]: ok done!

[the remark this answers is not in the document]
Comment (robmorgan) [reply .circleci/config.yml:21]: Yep, I've bumped them to more recent versions.

File: modules/run-consul/run-consul (hunk 6 above)
Comment (robmorgan) [reply modules/run-consul/run-consul:197]: Yeah I've stepped it back two levels.

File: .circleci/config.yml [hunk 12]
```diff
@@ -1,48 +1,57 @@
...
+            configure-environment-for-gruntwork-module \
+              --go-src-path ./test \
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply .circleci/config.yml:23]: Ok!

File: modules/consul-cluster/README.md [hunk 13]
```diff
@@ -118,12 +112,11 @@ Finally, you can try opening up the Consul UI in your browser at the URL `http:/
...
 ![Consul UI](https://github.com/hashicorp/terraform-google-consul/blob/master/_docs/consul-ui-screenshot.png?raw=true)
 
-
 ### Using the Consul agent on another Compute Instance
 
-The easiest way to run [Consul agent](https://www.consul.io/docs/agent/basics.html) and have it connect to the Consul 
+The easiest way to run [Consul agent](https://www.consul.io/docs/agent/basics.html) and have it connect to the Consul
 Server cluster is to note the [tag](https://cloud.google.com/compute/docs/vpc/add-remove-network-tags) used by a Compute
-Instance where the Consul agent is running, and specify it in the `allowed_inbound_tags_http_api` property of the 
+Instance where the Consul agent is running, and specify it in the `allowed_inbound_tags_http_api` property of the
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply modules/consul-cluster/README.md:140]: Hey @brikis98 I've undone a bunch of them in this file, but it looks like a lot of the lines were ending with excess whitespace, and unfortunately just about any editor I use at least tries to cleanup unnecessary space at the end of a line. Have another go trying to review this file. Just a heads up, I've left all my other 'prettier' changes in the other Markdown files as 1) I had to change quite a bit of them anyway 2) I think the prettier changes are less significant.

**Review by brikis98 — APPROVED** [review]
LGTM!

**Comment by robmorgan** [comment]
> LGTM!

Nice! Just realized I better write a migration guide for this one tomorrow.

**Pull request merged**